File: src/bot_control/scripts/reading_laser.cpp
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/laser_scan.hpp"

using std::placeholders::_1;

class LidarSub : public rclcpp::Node
{
public:
    LidarSub()
    : Node("lidar_sub_node")
    {
        // Subscriber to original /scan topic
        lidar_sub_ = this->create_subscription<sensor_msgs::msg::LaserScan>(
            "/scan",
            10,
            std::bind(&LidarSub::lidarCallback, this, _1)
        );

        // Publisher to new /filtered_scan topic
        filtered_pub_ = this->create_publisher<sensor_msgs::msg::LaserScan>(
            "/filtered_scan",
            10
        );
    }

private:
    void lidarCallback(const sensor_msgs::msg::LaserScan::SharedPtr msg)
    {
        // Filter from 0 to 120 degrees (out of full 360°)
        size_t total_samples = msg->ranges.size();
        size_t fov_samples = static_cast<size_t>(total_samples * (120.0 / 360.0));

        auto filtered_scan = sensor_msgs::msg::LaserScan();
        filtered_scan.header = msg->header;
        filtered_scan.angle_min = msg->angle_min;
        filtered_scan.angle_max = msg->angle_min + (msg->angle_increment * fov_samples);
        filtered_scan.angle_increment = msg->angle_increment;
        filtered_scan.time_increment = msg->time_increment;
        filtered_scan.scan_time = msg->scan_time;
        filtered_scan.range_min = msg->range_min;
        filtered_scan.range_max = msg->range_max;

        // Copy only the required ranges and intensities
        filtered_scan.ranges.assign(msg->ranges.begin(), msg->ranges.begin() + fov_samples);
        filtered_scan.intensities.assign(msg->intensities.begin(), msg->intensities.begin() + fov_samples);

        // Publish the filtered scan
        filtered_pub_->publish(filtered_scan);

        // (Optional) Also print the filtered ranges
        RCLCPP_INFO(this->get_logger(), "Filtered scan published with %zu samples.", fov_samples);
    }

    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr lidar_sub_;
    rclcpp::Publisher<sensor_msgs::msg::LaserScan>::SharedPtr filtered_pub_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<LidarSub>());
    rclcpp::shutdown();
    return 0;
}

File: src/bot_control/bot_control/reading_laser.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import LaserScan

class LidarSub(Node):

    def __init__(self):
        super().__init__("lidar_sub_node")

        # Subscriber to original /scan topic
        self.lidar_sub = self.create_subscription(
            LaserScan, 
            '/scan', 
            self.lidar_cb, 
            10
        )

        # Publisher to new /filtered_scan topic
        self.filtered_pub = self.create_publisher(
            LaserScan, 
            '/filtered_scan', 
            10
        )

    def lidar_cb(self, data):
        # Filter from 0 to 120 degrees (out of full 360°)
        total_samples = len(data.ranges)
        fov_samples = int(total_samples * (120.0 / 360.0))  # 120 degrees worth of samples
        
        filtered_scan = LaserScan()
        filtered_scan.header = data.header
        filtered_scan.angle_min = data.angle_min
        filtered_scan.angle_max = data.angle_min + (data.angle_increment * fov_samples)
        filtered_scan.angle_increment = data.angle_increment
        filtered_scan.time_increment = data.time_increment
        filtered_scan.scan_time = data.scan_time
        filtered_scan.range_min = data.range_min
        filtered_scan.range_max = data.range_max
        filtered_scan.ranges = data.ranges[0:fov_samples]
        filtered_scan.intensities = data.intensities[0:fov_samples]

        # Publish the filtered scan
        self.filtered_pub.publish(filtered_scan)

        # (Optional) Also print the filtered ranges
        self.get_logger().info(f"Filtered scan published with {fov_samples} samples.")

def main(args=None):
    rclpy.init(args=args)
    sensor_sub = LidarSub()
    rclpy.spin(sensor_sub)
    sensor_sub.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
